apps/expo/src/components/tracker/set-entry.tsx

```diff
@@ -1,31 +1,354 @@
 
 import { Feather } from '@expo/vector-icons'
+import { SetSchema } from '^/packages/validators/src'
+import { produce } from 'immer'
+import { z } from 'zod'
 
+import type { TExercise } from '~/components/tracker/workout-tracker'
+import Toggle from '~/components/toggle/toggle'
+import SetTypeButton from '~/components/tracker/set-type-button'
 import Button from '~/components/ui/button/button'
 
-const SetEntry: React.FC = () => {
+export type TSet = z.infer<typeof SetSchema>
+export interface ISetEntryProps {
+  set: TSet
+  exerciseIndex: number
+  workoutUpdater: React.Dispatch<React.SetStateAction<TExercise[]>>
+  setIndex: number
+}
+
+export interface IDeleteButtonProps {
+  handlePress: () => void
+}
+
+const DeleteButton: React.FC<IDeleteButtonProps> = ({ handlePress }) => {
   return (
-    <View className="flex flex-row items-center justify-between px-2">
-      <View className="flex basis-14 flex-row justify-start py-1" testID="set-type-button-container">
-        <Button value="W" className="h-9 w-full" color="dark" />
-      </View>
-      <View className="flex basis-24 items-center py-1" testID="unilateral-switch-container">
-        <Switch className="h-9" />
-      </View>
-      <View className="basis-20 py-1" testID="weight-input-container">
-        <TextInput className="h-9 rounded-lg bg-slate-900 text-center text-white" />
-      </View>
-      <View className="basis-20 py-1" testID="reps-input-container">
-        <TextInput className="h-9 rounded-lg bg-slate-900 text-center text-white" />
-      </View>
-      <View className="basis-12 py-1" testID="completed-button-container">
-        <Button className="flex h-9 items-center justify-center" color="dark">
-          <Feather name="check" color="white" size={19} />
-        </Button>
-      </View>
+    <View className='mx-2 flex w-12 justify-center'>
+      <Button
+        className='bg-light-red h-full w-full items-center justify-center rounded-lg'
+        onPress={handlePress}
+      >
+        <Feather name='trash-2' size={24} color='white' />
+      </Button>
     </View>
   )
 }
 
+const SetEntry: React.FC<ISetEntryProps> = ({ set, workoutUpdater, exerciseIndex, setIndex }) => {
+  const [isUnilateral, setIsUnilateral] = useState(set.unilateral)
+  const [completed, setCompleted] = useState(false)
+  const [inErrorState, setInErrorState] = useState({
+    weight: false,
+    reps: false,
+    leftWeight: false,
+    rightWeight: false,
+    leftReps: false,
+    rightReps: false,
+  })
+
+  const handleWeightChange = useCallback(
+    (e: NativeSyntheticEvent<TextInputChangeEventData>, idx: number) => {
+      workoutUpdater(
+        produce((draft) => {
+          const { sets } = draft[exerciseIndex]!
+          const setIndex = sets.findIndex((s) => s.id === set.id)
+          sets[setIndex]!.weight[idx] = Number(e.nativeEvent.text)
+        }),
+      )
+      setInErrorState(
+        produce((draft) => {
+          if (isUnilateral) {
```

Comment:
The anonymous function param to this `produce` call could be extracted to a named function, and then also used on line 95. You would just pass a function to the callback to set the `draft` properties that you need.

```ts
const produceCallback = (setLeft: function, setRight: function) => {
  // all the logic
  idx === 0 ? setLeft(true) : setRight(true);

  // and in the else block
  idx === 0 ? setLeft(false) : setRight(false);
};
```

The call to `produce` would look like:

```ts
produce(produceCallback((enabled) => draft.leftWeight = enabled, (enabled) => draft.rightWeight = enabled));

// and on line 95

produce(produceCallback((enabled) => draft.leftReps = enabled, (enabled) => draft.rightReps = enabled));
```

Although I understand if you don't want to try this. It's a more advanced JS/TS technique called "closures" that took me a minute to wrap my head around. Here's a ChatGPT explanation: https://chat.openai.com/share/2f031d05-f348-4802-802c-9dc8eb68114a

Reply:
Oh thats neat I did not know about this, will def implement this change. I am aware that this pr has a lot of duplicate and refactoring possibilities so sorry about that 💀 im trying to get as much done as possible rn